components/nexus-security/src/main/java/org/sonatype/security/realms/RolePermissionResolverImpl.java

```diff
@@ -109,35 +127,66 @@ protected void resolvePermissionsInRole(final String roleString, final Collectio
 
           // process the roles this role has recursively
           rolesToProcess.addAll(role.getRoles());
+
           // add the permissions this role has
           for (String privilegeId : role.getPrivileges()) {
-            Set<Permission> set = getPermissions(privilegeId);
-            permissions.addAll(set);
+            Permission permission = permission(privilegeId);
+            if (permission != null) {
+              permissions.add(permission);
+            }
           }
         }
         catch (NoSuchRoleException e) {
-          // skip
+          log.trace("Ignoring missing role: {}", roleId, e);
         }
       }
     }
 
     // cache result of (non-trivial) computation
-    permissionsCache.put(roleString, permissions);
+    rolePermissionsCache.put(roleString, permissions);
+
+    return permissions;
   }
 
-  protected Set<Permission> getPermissions(final String privilegeId) {
-    try {
-      final CPrivilege privilege = configuration.readPrivilege(privilegeId);
-      for (PrivilegeDescriptor descriptor : privilegeDescriptors) {
-        final String permission = descriptor.buildPermission(privilege);
-        if (permission != null) {
-          return Collections.singleton(permissionFactory.create(permission));
-        }
+  /**
+   * Returns the descriptor for the given privilege-type or {@code null}.
+   */
+  @Nullable
+  private PrivilegeDescriptor descriptor(final String privilegeType) {
+    assert privilegeType != null;
```

Comment:
because its private method, assume caller makes sane, to avoid byte-code overhead